File: code/sorting/bead_sort/bead_sort.cs
/* Part of Cosmos by OpenGenus Foundation */

using System;
using System.Linq;

namespace CS
{
    public class BeadSort
    {
        public BeadSort()
        {
        }

        static int[] Sort(int[] arr) {
            int max = arr.Max();

            var grid = new bool[arr.Length, max];

            int[] levelCount = new int[max];
            
            for(int i = 0; i < max; i++) {
                levelCount[i] = 0;
                for(int j = 0; j < arr.Length; j++) {
                    grid[j, i] = false;
                }
            }

            foreach (var num in arr)
            {
                var tmpNum = num;
                for (int j = 0; tmpNum > 0; j++, tmpNum--)
                {
                    grid[levelCount[j]++, j] = true;
                }
            }

            var sorted = new int[arr.Length];
            for (int i = 0; i < arr.Length; i++)
            {
                int putt = 0;
                for (int j = 0; j < max && grid[arr.Length - 1 - i, j] == true; j++)
                {
                    putt++;
                }
                sorted[i] = putt;
            }

            return sorted;

        }

        static void Main(string[] args)
        {
            const int numbers = 25;
            int[] arr = new int[numbers];

            var rand = new Random();
            for (int i = 0; i < numbers; i++)
            {
                arr[i] = rand.Next(99);
            }


            Console.WriteLine("Before Sorting");
            foreach (var num in arr)
            {
                Console.Write(num + " ");
            }

            var sorted = Sort(arr);
            Console.WriteLine("\nSorted:");
            foreach (var num in sorted)
            {
                Console.Write(num + " ");
            }
        }
    }
}

File: cosmos-master/code/sorting/bead_sort/BeadSort.java
import java.util.Arrays;
// Part of Cosmos by OpenGenus Foundation
public class BeadSort {

    private enum BeadSortStatus {
        MARKED,
        NOT_MARKED,
    }

    public static void main(String[] args) {
        int[] arrayToSort = new int[]{4, 1, 6, 2, 40, 5, 3, 8, 7};
        System.out.println(Arrays.toString(beadSort(arrayToSort)));
    }

    public static int[] beadSort(int[] arr) {
        int max = 0;
        for (int anArr : arr) {
            if (anArr > max) {
                max = anArr;
            }
        }

        //Set up abacus
        BeadSortStatus[][] grid = new BeadSortStatus[arr.length][max];

        int[] levelcount = new int[max];
        for(int i = 0; i < max; i++) {
            levelcount[i] = 0;
            for(int j = 0; j < arr.length; j++) {
                grid[j][i] = BeadSortStatus.NOT_MARKED;
            }
        }

        //Drop the beads
        for (int anArr : arr) {
            int num = anArr;
            for (int j = 0; num > 0; j++, num--) {
                grid[levelcount[j]++][j] = BeadSortStatus.MARKED;
            }
        }

        //Count the beads
        int[] sorted=new int[arr.length];
        for(int i = 0; i < arr.length; i++) {
            int putt = 0;
            for(int j = 0; j < max && grid[arr.length - 1 - i][j] == BeadSortStatus.MARKED; j++) {
                putt++;
            }
            sorted[i] = putt;
        }

        return sorted;
    }

}
